coordinator/ts/proof/proof.service.ts

```diff
@@ -0,0 +1,144 @@
+    this.deployment.setHre(hre);
+    this.cryptoService = CryptoService.getInstance();
+  }
+
+  /**
+   * Generate proofs for message processing and tally
+   *
+   * @param args - generate proofs arguments
+   * @returns - generated proofs for message processing and tally
+   */
+  async generate({
+    poll,
+    maciContractAddress,
+    tallyContractAddress,
+    useQuadraticVoting,
+    encryptedCoordinatorPrivateKey,
+    startBlock,
+    endBlock,
+    blocksPerBatch,
+  }: IGenerateArgs): Promise<IGenerateData> {
+    const maciContract = await this.deployment.getContract<MACI>({
+      name: EContracts.MACI,
+      address: maciContractAddress,
+    });
+
+    const signer = await this.deployment.getDeployer();
+    const pollAddress = await maciContract.polls(poll);
+    const pollContract = await this.deployment.getContract<Poll>({ name: EContracts.Poll, address: pollAddress });
+    const { messageAq: messageAqAddress } = await pollContract.extContracts();
+    const messageAq = await this.deployment.getContract<AccQueue>({
+      name: EContracts.AccQueue,
+      address: messageAqAddress,
+    });
+
+    const isStateAqMerged = await pollContract.stateAqMerged();
+
+    if (!isStateAqMerged) {
+      throw new Error("The state tree has not been merged yet. Please use the mergeSignups subcommmand to do so.");
+    }
+
+    const messageTreeDepth = await pollContract.treeDepths().then((depths) => Number(depths[2]));
+
+    const mainRoot = await messageAq.getMainRoot(messageTreeDepth.toString());
+
+    if (mainRoot.toString() === "0") {
+      throw new Error("The message tree has not been merged yet. Please use the mergeMessages subcommmand to do so.");
+    }
+
+    const privateKey = await fs.promises.readFile(path.resolve(process.env.COORDINATOR_PRIVATE_KEY_PATH!));
+    const maciPrivateKey = PrivKey.deserialize(this.cryptoService.decrypt(privateKey, encryptedCoordinatorPrivateKey));
+    const coordinatorKeypair = new Keypair(maciPrivateKey);
+
+    const maciState = await ProofGenerator.prepareState({
+      maciContract,
+      pollContract,
+      messageAq,
+      maciPrivateKey,
+      coordinatorKeypair,
+      pollId: poll,
+      signer,
+      options: {
+        startBlock,
+        endBlock,
+        blocksPerBatch,
+      },
+    });
+
+    const foundPoll = maciState.polls.get(BigInt(poll));
+
+    if (!foundPoll) {
+      throw new Error(`Poll ${poll} not found`);
+    }
+
+    const proofGenerator = new ProofGenerator({
+      poll: foundPoll,
+      maciContractAddress,
+      tallyContractAddress,
+      tally: this.getZkeyFiles(process.env.COORDINATOR_TALLY_ZKEY_NAME!, useQuadraticVoting),
+      mp: this.getZkeyFiles(process.env.COORDINATOR_MESSAGE_PROCESS_ZKEY_NAME!, useQuadraticVoting),
+      outputDir: path.resolve("./proofs"),
+      tallyOutputFile: path.resolve("./tally"),
+      useQuadraticVoting,
+    });
+
+    const [processProofs, tallyProofs] = await Promise.all([
```

Comment:
I'm afraid these need to be sequential, can't tally until all messages have been processed. Though we could run full local message processing (in batches as normal), with concurrent proof generation for it (as each batch is ready); and once the messages are processed locally (proofs might not have finished yet) in `Poll`, we can start tallying in `Poll` + as batches ready, start proof generation for tally

Reply:
Proof parallel processing will be implemented later, but good to know that I need to process them sequentially.